File: java/TestFluxVM.java
public class TestFluxVM {
    static int passed = 0, failed = 0;
    
    static void check(String name, boolean cond) {
        if (cond) { System.out.printf("  %-50sPASS%n", name); passed++; }
        else { System.out.printf("  %-50sFAIL%n", name); failed++; }
    }
    
    static byte[] bc(int... vals) {
        byte[] r = new byte[vals.length];
        for (int i = 0; i < vals.length; i++) r[i] = (byte)vals[i];
        return r;
    }

    public static void main(String[] args) {
        System.out.println("\nFLUX Java VM Tests\n==================\n");

        { FluxVM v = new FluxVM(); v.execute(bc(0x00)); check("halt", v.halted && v.cycles==1); }
        { FluxVM v = new FluxVM(); v.execute(bc(0x01,0x00)); check("nop", v.cycles==2); }
        { FluxVM v = new FluxVM(); v.execute(bc(0x18,0,42,0x00)); check("movi", v.gp[0]==42); }
        { FluxVM v = new FluxVM(); v.execute(bc(0x18,0,0xFB,0x00)); check("movi_neg", v.gp[0]==-5); }
        { FluxVM v = new FluxVM(); v.execute(bc(0x40,0,0xE8,0x03,0x00)); check("movi16", v.gp[0]==1000); }
        { FluxVM v = new FluxVM(); v.execute(bc(0x40,0,0x00,0x80,0x00)); check("movi16_neg", v.gp[0]==-32768); }
        { FluxVM v = new FluxVM(); v.execute(bc(0x18,0,10,0x18,1,20,0x20,2,0,1,0x00)); check("add", v.gp[2]==30); }
        { FluxVM v = new FluxVM(); v.execute(bc(0x18,0,7,0x18,1,6,0x22,2,0,1,0x00)); check("mul", v.gp[2]==42); }
        { FluxVM v = new FluxVM(); v.execute(bc(0x18,0,42,0x18,1,7,0x23,2,0,1,0x00)); check("div", v.gp[2]==6); }
        { FluxVM v = new FluxVM(); v.execute(bc(0x18,0,10,0x08,0,0x08,0,0x09,0,0x00)); check("inc_dec", v.gp[0]==11); }
        { FluxVM v = new FluxVM(); v.execute(bc(0x18,0,42,0x0C,0,0x18,0,0,0x0D,1,0x00)); check("push_pop", v.gp[1]==42); }
        { FluxVM v = new FluxVM(); v.execute(bc(0x18,0,3,0x18,1,7,0x2A,2,0,1,0x2B,3,0,1,0x00)); check("min_max", v.gp[2]==3 && v.gp[3]==7); }
        { FluxVM v = new FluxVM(); v.execute(bc(0x18,0,5,0x18,1,0,0x08,1,0x09,0,0x3D,0,0xFC,0,0x00)); check("jnz_loop", v.gp[0]==0 && v.gp[1]==5); }
        { FluxVM v = new FluxVM(); // fibonacci
            v.execute(bc(0x18,0,1,0x18,1,1,0x18,2,10, 0x20,3,0,1,0x3A,0,1,0,0x3A,1,3,0,0x09,2, 0x3D,2,0xF2,0, 0x00));
            check("fibonacci", v.gp[1]==144);
        }

        System.out.printf("%n==================%nResults: %d passed, %d failed%n", passed, failed);
        if (failed > 0) System.exit(1);
    }
}

File: cpp/test_flux_vm.cpp
#include "flux_vm.hpp"
#include <cassert>
#include <cstdio>
#include <vector>

static int passed = 0, failed = 0;
#define TEST(name) printf("  %-50s", #name);
#define PASS() do { printf("PASS\n"); passed++; } while(0)
#define ASSERT_EQ(a, b) do { if ((a) != (b)) { printf("FAIL: %s=%lld, expected %lld\n", #a, (long long)(a), (long long)(b)); failed++; return; } } while(0)

void test_halt() { TEST(halt); FluxVM vm; vm.execute({OP_HALT}); assert(vm.halted); assert(vm.cycles==1); PASS(); }
void test_nop() { TEST(nop); FluxVM vm; vm.execute({OP_NOP, OP_HALT}); assert(vm.cycles==2); PASS(); }
void test_movi() { TEST(movi); FluxVM vm; vm.execute({OP_MOVI, 0, 42, OP_HALT}); assert(vm.gp[0]==42); PASS(); }
void test_movi_neg() { TEST(movi_neg); FluxVM vm; vm.execute({OP_MOVI, 0, (uint8_t)-5, OP_HALT}); assert(vm.gp[0]==-5); PASS(); }
void test_movi16() { TEST(movi16); FluxVM vm; vm.execute({OP_MOVI16, 0, 0xE8, 0x03, OP_HALT}); assert(vm.gp[0]==1000); PASS(); }
void test_movi16_neg() { TEST(movi16_neg); FluxVM vm; vm.execute({OP_MOVI16, 0, 0x00, 0x80, OP_HALT}); assert(vm.gp[0]==-32768); PASS(); }
void test_add() { TEST(add); FluxVM vm; vm.execute({OP_MOVI,0,10, OP_MOVI,1,20, OP_ADD,2,0,1, OP_HALT}); assert(vm.gp[2]==30); PASS(); }
void test_mul() { TEST(mul); FluxVM vm; vm.execute({OP_MOVI,0,7, OP_MOVI,1,6, OP_MUL,2,0,1, OP_HALT}); assert(vm.gp[2]==42); PASS(); }
void test_div() { TEST(div); FluxVM vm; vm.execute({OP_MOVI,0,42, OP_MOVI,1,7, OP_DIV,2,0,1, OP_HALT}); assert(vm.gp[2]==6); PASS(); }
void test_inc_dec() { TEST(inc_dec); FluxVM vm; vm.execute({OP_MOVI,0,10, OP_INC,0, OP_INC,0, OP_DEC,0, OP_HALT}); assert(vm.gp[0]==11); PASS(); }
void test_push_pop() { TEST(push_pop); FluxVM vm; vm.execute({OP_MOVI,0,42, OP_PUSH,0, OP_MOVI,0,0, OP_POP,1, OP_HALT}); assert(vm.gp[1]==42); PASS(); }
void test_cmp_eq() { TEST(cmp_eq); FluxVM vm; vm.execute({OP_MOVI,0,5, OP_MOVI,1,5, OP_CMP_EQ,2,0,1, OP_HALT}); assert(vm.gp[2]==1); PASS(); }
void test_min_max() { TEST(min_max); FluxVM vm; vm.execute({OP_MOVI,0,3, OP_MOVI,1,7, OP_MIN,2,0,1, OP_MAX,3,0,1, OP_HALT}); assert(vm.gp[2]==3); assert(vm.gp[3]==7); PASS(); }
void test_jnz_loop() {
    TEST(jnz_loop);
    FluxVM vm;
    vm.execute({OP_MOVI,0,5, OP_MOVI,1,0, OP_INC,1, OP_DEC,0, OP_JNZ,0,(uint8_t)-4,0, OP_HALT});
    assert(vm.gp[0]==0); assert(vm.gp[1]==5); PASS();
}
void test_fibonacci() {
    TEST(fibonacci);
    FluxVM vm;
    vm.execute({
        OP_MOVI,0,1, OP_MOVI,1,1, OP_MOVI,2,10,
        OP_ADD,3,0,1, OP_MOV,0,1,0, OP_MOV,1,3,0, OP_DEC,2,
        OP_JNZ,2,(uint8_t)-14,0, OP_HALT
    });
    assert(vm.gp[1]==144); PASS();
}

int main() {
    printf("\nFLUX C++ VM Tests\n==================\n\n");
    test_halt(); test_nop(); test_movi(); test_movi_neg(); test_movi16();
    test_movi16_neg(); test_add(); test_mul(); test_div(); test_inc_dec();
    test_push_pop(); test_cmp_eq(); test_min_max(); test_jnz_loop(); test_fibonacci();
    printf("\n==================\nResults: %d passed, %d failed\n", passed, failed);
    return failed;
}
